instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/TripSessionsBillingTest.kt

```diff
@@ -0,0 +1,316 @@
+import com.mapbox.navigation.instrumentation_tests.utils.routes.RoutesProvider.toNavigationRoutes
+import com.mapbox.navigation.testing.ui.BaseCoreNoCleanUpTest
+import com.mapbox.navigation.testing.ui.utils.getMapboxAccessTokenFromResources
+import com.mapbox.navigation.testing.ui.utils.runOnMainSync
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.filter
+import kotlinx.coroutines.flow.take
+import org.junit.After
+import org.junit.Assert.assertEquals
+import org.junit.Assert.assertNotNull
+import org.junit.Assert.assertTrue
+import org.junit.Assert.fail
+import org.junit.Before
+import org.junit.Rule
+import org.junit.Test
+import java.net.URI
+import java.util.concurrent.TimeUnit
+
+class TripSessionsBillingTest : BaseCoreNoCleanUpTest() {
+
+    @get:Rule
+    val mapboxNavigationRule = MapboxNavigationRule()
+
+    private lateinit var mapboxNavigation: MapboxNavigation
+
+    private val route = RoutesProvider.dc_very_short(context).toNavigationRoutes()[0]
+
+    private lateinit var httpEventsObserver: HttpServiceEventsObserver
+
+    override fun setupMockLocation(): Location = mockLocationUpdatesRule.generateLocationUpdate {
+        route.routeOptions.coordinatesList().first().run {
+            latitude = latitude()
+            longitude = longitude()
+        }
+        bearing = 190f
+    }
+
+    @Before
+    fun setup() {
+        httpEventsObserver = HttpServiceEventsObserver.register()
+
+        runOnMainSync {
+            mapboxNavigation = MapboxNavigationProvider.create(
+                NavigationOptions.Builder(context)
+                    .accessToken(getMapboxAccessTokenFromResources(context))
+                    .routingTilesOptions(
+                        RoutingTilesOptions.Builder()
+                            .tilesBaseUri(URI(mockWebServerRule.baseUrl))
+                            .build()
+                    )
+                    .build()
+            )
+        }
+    }
+
+    @After
+    fun cleanUp() {
+        HttpServiceFactory.getInstance().setInterceptor(null)
+    }
+
+    /**
+     * Waits for exactly [count] elements that match [predicate] to appear
+     * in [HttpServiceEventsObserver] to avoid tests flakiness,
+     * than waits for 1 second for more http events to arrive.
+     * This is made in order to catch bugs
+     * where more than expected billing events sent to the server.
+     *
+     * @return All billing [HttpServiceEvent.Request]'s appeared within waiting time.
+     * Number of returned items can be more that [count].
+     * Returned elements might not match the [predicate].
+     */
+    private suspend fun waitForBillingEvents(
+        count: Int,
+        predicate: suspend (HttpServiceEvent) -> Boolean = { true }
+    ): List<HttpServiceEvent.Request> {
+        httpEventsObserver
+            .billingRequestsFlow()
+            .filter(predicate)
+            .take(count)
```

Comment:
Now when there are no events, we'll get TimeoutException without the stacktrace leading to the line which failed.
Can we do sth like this:
1. Change `waitForBillingEvents` into:
```
    private suspend fun waitForBillingEvents(
        count: Int,
        description: String = "",
        predicate: suspend (HttpServiceEvent) -> Boolean = { true }
    ): List<HttpServiceEvent.Request> {
        try {
            httpEventsObserver
                .billingRequestsFlow()
                .filter(predicate)
                .take(count)
                .toList()
        } catch (ex: CancellationException) {
            Log.e("[WaitForBillingEvents]", "Cancelled waiting for $count $description events")
            throw ex
        }

        delay(1000)
        return httpEventsObserver.billingRequests()
    }
```    
and pass the description in tests (possible values are: "FreeDrive", "ActiveGuidance", "any").
This way we can at least use logs to see what exactly has gone wrong.
Or any other idea to make it more clear that you might have.

Reply:
Good, idea, thank you! We can fail with detailed message right in the catch block: `fail("Timed out waiting for $count $description event(s)")`